keps/sig-network/3453-minimize-iptables-restore/README.md

```diff
@@ -0,0 +1,468 @@
+
+## Motivation
+
+### Goals
+
+- Improve the performance of kube-proxy in iptables mode, by optimizing
+  its calls to `iptables-restore` by not including rules that haven't
+  changed.
+
+### Non-Goals
+
+- Everything else.
+
+## Proposal
+
+### General Plan
+
+`iptables-restore` has the constraint that if you want to specify
+_any_ rule in a given iptables chain, then you must specify _every_
+rule in that chain. However, if you want to leave a chain completely
+unchanged, then you can just fail to mention any rules in that chain,
+and it will be left untouched. (If you want to delete a chain you have
+to explicitly request that.)
+
+Thus, [the proposed PR] keeps track of which Service and EndpointSlice
+objects have changed since the last `iptables-restore`, and only
+outputs the Service-specific chains for a Service (the `KUBE-SVC-*`,
+`KUBE-SVL-*`, `KUBE-EXT-*`, `KUBE-FW-*`, and `KUBE-SEP-*` chains) if
+the Service or its EndpointSlice(s) changed since the last
+`iptables-restore`. (The `KUBE-SERVICES`, `KUBE-EXTERNAL-SERVICES`,
+and `KUBE-NODEPORTS` chains are written in their entirety on every
+sync.)
+
+In theory this could be optimized further; eg, there are situations
+where only `KUBE-FW-*` changes, or where we need to rewrite the
+`KUBE-SERVICES` rules for a service but not any of the
+service-specific chains, or vice versa. However, these other
+optimizations would be more complicated, and would only result in a
+small additional performance improvement relative to the main
+improvement of not rewriting _every_ service and endpoint rule on
+_every_ resync.
+
+In the event that a partial `iptables-restore` call fails, kube-proxy
+will queue another resync, and fall back to doing a full
+`iptables-restore` that time. (Note that there are no expected cases
+where a partial restore would fail but a full restore would succeed;
+this would always indicate a bug in the code.) Likewise, it always
+does a full resync when there are topology-related changes to Node
+labels, and it always does a full resync at least once every
+`iptablesSyncPeriod`.
+
+[the proposed PR]: https://github.com/kubernetes/kubernetes/pull/110268
+
+### Risks and Mitigations
+
+#### Catastrophic Failure
+
+In the event of bugs in the code, kube-proxy could fail
+catastrophically, entirely breaking the cluster. This would be "bad".
+
+#### Subtle Optimization Failure
+
+The code could also fail more subtly. In the worst case, every attempt
+to do a partial restore might fail, causing the code to fall back
+to the slower full `iptables-restore` on the next sync, with the end
+result that kube-proxy becomes _less_ efficient than the old code that
+would have just done a full restore the first time.
+
+Less obviously, the new code might reliably cause a fail-and-retry
+only in certain circumstances (eg, when a service goes from 1 endpoint
+to 0, or when a particular Service property changes).
+
+Since, as mentioned above, there are no _expected_ cases where a
```

Comment:
it can fail because of external things, it is not common but it can happen, at the end is an userspace application on the sysmte

Reply:
oh, you mean, to fail just the partial restore, ...

Reply:
what makes an iptables-restore to fail exactly? adding wrong rules will not make it fail, right?
I don't know if this can be meaningful, the errors we want to detect are only if the stored rule are not the right ones ...

Reply:
I don't know what would make it fail other than a) syntax errors (GIGO, deterministic); or b) xtable locked (never actually seen this, AFAIK).

Reply:
that is my point, the metric here will not give us much feedback, a) must not happen b) can happen in any execution

Reply:
what we want to know if we screwed the service with the partial restore, but kube-proxy doesn't have those insights, I don't feel this metric will be useful

Reply:
The only way I see we can do to know that the feature is ok is using one e2e test that generate/update Services incrementally and assert they are correct

Reply:
`iptables-restore` will fail if the result of applying the restore would result in an inconsistent state. Specifically:
- if you add a rule with a `-j CHAIN` to a chain that doesn't exist (and which you aren't creating in the restore)
- if you delete a chain which is jumped to by an existing rule (that you aren't also deleting in the same restore)

Existing kube-proxy never hits either of those cases(*) because it always writes out every single rule and chain that kube-proxy cares about. But if the partial restore generation code is buggy, then it could potentially generate inconsistent partial restores, which would fail, but then a following full restore would succeed. (But this would only happen if the partial restore code was buggy, which is what I meant about there being no expected cases of it.)

(*) before 1.25 it could happen if KUBE-MARK-DROP was missing

Reply:
oh, that makes sense then

Reply:
```
cat save.txt 
# Generated by iptables-save v1.8.7 on Wed Sep 14 11:48:16 2022
*filter
:INPUT ACCEPT [0:0]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
-A INPUT -i eth0 -p tcp -m tcp --dport 22 -j ACCEPT
-A NOTEXIST -i eth0 -p tcp -m tcp --dport 22 -j ACCEPT
COMMIT
# Completed on Wed Sep 14 11:48:16 2022
# iptables-restore < save.txt 
iptables-restore: line 7 failed
# echo $?
1
```

Reply:
you are completely right, this metric will be very useful in this case